### ISSUE TITLE
[Feature Research]: Image Captioning Models

### ISSUE THREAD

**Issue by cobycloud** [swarmauri/swarmauri-sdk#1353]
### Feature Name

Image Captioning Models

### Feature Description

Research state‑of‑the‑art deep learning architectures for automatically generating descriptive captions from static images.

### Research Findings

Investigate classical encoder–decoder approaches (Show & Tell, Show, Attend & Tell), and modern transformer‑based vision‑language models (BLIP, OFA, Flamingo). Evaluate open‑source implementations, COCO/Flickr30k benchmark scores (BLEU, METEOR, CIDEr), pretrained weight availability, inference latency, and model size. Investigate SOTA solutions.

### Potential Impact

Enables automatic alt‑text generation for accessibility, metadata creation for search/indexing, and rapid content tagging in Swarmauri applications.

### Additional Resources (optional)

Vinyals et al., “Show and Tell: A Neural Image Caption Generator” (2015)

Li et al., “BLIP: Bootstrapping Language-Image Pre-training” (2022)

GitHub: Salesforce/BLIP, Salesforce/OFA

### Feature Priority

Medium

**Comment by Rexon-Pambujya** [comment]
Hi! I’d like to work on this issue.

Before I start, I wanted to clarify the expected deliverable so I can align with the project’s needs:

1. Are you expecting primarily a **research/documentation contribution** (comparison of captioning architectures, benchmarks, implementation recommendations), or an actual **code implementation/prototype**?

2. If implementation is preferred, should it be:

* inference wrappers for existing pretrained models (BLIP/OFA/etc.)
* benchmarking scripts
* or integration into an existing module?

3. Is there a preferred location/format for this contribution (docs page, markdown report, notebook, module under a specific directory)?

4. Should the focus be on identifying the strongest open-source production-ready model for integration, or a broader comparative analysis of captioning approaches?

I’ve already started reviewing BLIP, OFA, OpenFlamingo, and classical encoder-decoder approaches, so I can proceed once I know the expected scope.

**Comment by cobycloud** [comment]
Research triage update — RETAIN, with primary scope under swarmauri_vlm_* rather than a new core captioning abstraction. Image captioning is now generally an image-text-to-text VLM workflow: https://huggingface.co/docs/transformers/en/tasks/image_text_to_text. Acceptance must define image input types, prompt behavior, batch/async support, returned message/text schema, model licensing, memory/latency, and deterministic fixtures. Next step: select one maintained runtime/provider and propose its standalone package.

**Comment by cobycloud** [comment]
## First-pass research — image captioning models

### Scope and repository fit

This ticket remains viable, but the first pass indicates that **image captioning should be implemented as a VLM workflow**, not as a second caption-only interface family.

The repository already has `IPredictVision` / `VLMBase` and provider VLM packages. The current interface is very loose (`*args`, `**kwargs`, and `any`), so a provider implementation should be paired with a small contract-hardening pass rather than introducing `IImageCaption` and `ImageCaptionBase`.

This matches the current Transformers taxonomy: image-text-to-text models are general VLMs that cover captioning, VQA, and related image-conditioned generation; task-specific image-to-text remains useful for narrower runtimes.

### Candidate matrix

| Candidate | Role | License/deployment | First-pass verdict |
|---|---|---|---|
| Gemini image understanding | Managed general-purpose VLM; captioning, classification, VQA, detection, and segmentation | Commercial API; supports URL, inline bytes, and uploaded files | **Recommended first managed provider** as `swarmauri_vlm_gemini`; it fills a missing VLM provider without inventing a caption-only class |
| Florence-2 large/base | Local task-focused vision-language runtime; captioning, detailed captioning, OCR, grounding | MIT model card; Transformers integration currently requires careful handling of `trust_remote_code` | **Recommended first specialized local provider**, subject to pinned revision and remote-code policy |
| Qwen2.5-VL 7B (and smaller maintained variants) | Local/server general-purpose VLM | Apache-2.0; Transformers and vLLM support | **Strong general local VLM candidate**, likely `swarmauri_vlm_qwen`; heavier than Florence for caption-only use |
| BLIP-2 OPT 2.7B | Established captioning/VQA baseline | MIT; roughly 7.2 GB FP16 weights per the model card, with quantized options | **Benchmark/regression baseline**, not the preferred first modern implementation |

### Proposed classes and contracts

No new image-caption base is proposed in this pass.

| Type | Package | Responsibility |
|---|---|---|
| `IPredictVision` | `swarmauri_core.vlms` | Retain as the shared VLM capability, but replace untyped signatures with explicit image/prompt, async, and batch semantics |
| `VLMBase` | `swarmauri_base.vlms` | Retain model validation/serialization boundary |
| `GeminiVLM` | `swarmauri_vlm_gemini` | Managed image + prompt to text; normalize file/URL/bytes inputs and keep credentials/upload handles out of serialized state |
| `FlorenceVLM` | `swarmauri_vlm_florence` | Local task-focused captions, detailed captions, OCR, and grounding; expose supported task prompts explicitly |
| `QwenVLM` | `swarmauri_vlm_qwen` | General image-conditioned generation and reasoning |

A caption should therefore look like a normal VLM call, for example:

```python
caption = model.predict_vision(
    image="fixture.jpg",
    prompt="Describe the image in one factual sentence.",
)
```

The shared interface should not add `stream`/`astream` merely because a hosted model can stream output tokens. Caption completion is still one logical result; provider-specific token iteration can be added only after the common streaming semantics are defined.

### Contract questions for the next pass

1. Normalize accepted image sources: local path, HTTPS URL, bytes, and optionally PIL image.
2. Define the return type as generated text (or the existing conversation/message type if VLMs are conversational), not provider response dictionaries.
3. Specify native async and ordered batch behavior.
4. Define EXIF orientation, RGBA conversion, MIME validation, maximum-size, and multi-image behavior.
5. Pin local model revisions and decide whether `trust_remote_code` is prohibited, opt-in, or restricted to audited revisions.
6. Keep credentials, processors, loaded weights, device handles, and uploaded-file handles out of serialized component state.

### First implementation acceptance criteria

- Deterministic synthetic fixtures for a simple scene, text-bearing image, rotated image, RGBA image, and malformed input.
- Short-caption and detailed-caption prompt profiles with factuality/hallucination checks.
- Sync, async, and ordered batch parity.
- Explicit optional dependencies and no import-time model download.
- Local checkpoint revision/checksum recorded; model license documented.
- Managed-provider tests use mocks by default and opt-in live integration tests.
- Errors normalized for invalid media, unsupported MIME type, provider rejection, timeout, and local OOM.

### Primary sources

- [Transformers image-text-to-text task guide](https://huggingface.co/docs/transformers/main/tasks/image_text_to_text)
- [Gemini image understanding](https://ai.google.dev/gemini-api/docs/image-understanding)
- [Microsoft Florence-2 large model card](https://huggingface.co/microsoft/Florence-2-large)
- [Qwen2.5-VL-7B-Instruct model card](https://huggingface.co/Qwen/Qwen2.5-VL-7B-Instruct)
- [Salesforce BLIP-2 OPT 2.7B model card](https://huggingface.co/Salesforce/blip2-opt-2.7b)

### First-pass disposition

**Retain and leave open.** Next research should compare Gemini, Florence-2, and Qwen on the same five fixtures, then select one managed and one local provider. This comment does not close or mark the ticket implemented.

**Comment by cobycloud** [comment]
## Follow-up research: canonical VLM portfolio and migration direction

Image captioning should be treated as one capability of the existing VLM family, not as a separate `IImageCaption` abstraction. Modern image-text-to-text models cover captioning alongside visual question answering, OCR, document/screenshot understanding, classification, detection, grounding, and sometimes segmentation.

### Current Swarmauri state

| Implementation | Current shape | Recommendation |
|---|---|---|
| `GroqVLM` | `VLMBase`; conversation, sync/async, streaming, batch | Keep and migrate to standalone `swarmauri_vlm_groq` |
| `HyperbolicVLM` | `VLMBase`; conversation, local-file encoding, streaming | Keep conditionally; repair defects and verify current provider conformance |
| `FalVLM` | `VLMBase`; direct image/prompt plus queued polling | Replace the current catch-all design; the listed OCR, captioning, moderation, image, and video endpoints do not share one stable request/response schema |
| `GroqVisionModel` | Legacy `LLMBase` duplicate | Deprecate/remove after migration |
| `HyperbolicVisionModel` | Legacy `LLMBase` duplicate | Deprecate/remove after migration |
| `FalAIVisionModel` | Legacy `LLMBase` duplicate | Deprecate/remove after migration |

The current `IPredictVision` contract is also too loose: its methods use untyped `*args/**kwargs`, while `VLMBase` mainly provides model-name validation. Before adding more providers, standardize multimodal conversation content, image sources, normalized output behavior, capability metadata, and error semantics.

### Proposed shared types

| Type | Package | Responsibility |
|---|---|---|
| `IPredictVision` | `swarmauri_core` | Typed multimodal conversation contract for `predict_vision`, `apredict_vision`, `batch`, and `abatch` |
| `IStreamingVLM` | `swarmauri_core` | Optional `stream`/`astream` capability; do not put unsupported streaming methods on every VLM |
| `VLMBase` | `swarmauri_base` | Registration, serialization, model validation, and capability metadata |
| `OpenAICompatibleVLM` | `swarmauri_base` | Reusable chat-completions transport, multimodal message formatting, retry behavior, and provider hooks |
| `TransformersVLM` | `swarmauri_vlm_transformers` | Generic local runtime for conformant `AutoModelForImageTextToText` models |

The canonical input should be conversation-first because image understanding commonly needs multiple images and follow-up turns. Message content should normalize at least text, image URL, base64/data URL, local path/upload, MIME type, and optional detail/resolution. A simple `caption(image, prompt)` helper may be provided as a convenience over that contract.

### Proposed concrete classes

| Class | Inheritance | Package | Priority | Role |
|---|---|---|---:|---|
| `GeminiVLM` | `VLMBase, IStreamingVLM` | `swarmauri_vlm_gemini` | P1 | First broad managed implementation; general image understanding |
| `OpenAIVLM` | `VLMBase, IStreamingVLM` | `swarmauri_vlm_openai` | P1 | Responses API, file IDs, image detail, structured output |
| `AnthropicVLM` | `VLMBase, IStreamingVLM` | `swarmauri_vlm_anthropic` | P1 | Claude Messages image blocks, file references, multi-image conversation |
| `GroqVLM` | `OpenAICompatibleVLM` | `swarmauri_vlm_groq` | P1 | Preserve existing low-latency hosted behavior through the canonical base |
| `TransformersVLM` | `VLMBase` | `swarmauri_vlm_transformers` | P1 | General local/private inference |
| `FlorenceVLM` | `VLMBase` | `swarmauri_vlm_florence` | P1 | Specialized caption, OCR, detection, grounding, and segmentation results |
| `MistralVLM` | `OpenAICompatibleVLM` | `swarmauri_vlm_mistral` | P2 | Managed Mistral vision models |
| `HyperbolicVLM` | `OpenAICompatibleVLM` | `swarmauri_vlm_hyperbolic` | P2 | Hosted open VLMs, subject to live conformance |
| `QwenVLM` | `TransformersVLM` | `swarmauri_vlm_qwen` | P2 | Qwen-specific preprocessing, video, grounding, or runtime extensions |
| `FalAIFlorenceVLM` | `VLMBase` | `swarmauri_vlm_falai` | P2 | Validated Florence-specific Fal schema |
| `FalAIMoondreamVLM` | `VLMBase` | `swarmauri_vlm_falai` | P3 | Validated lightweight caption/VQA endpoint |

`SmolVLM` should initially be selected through `TransformersVLM`, not receive a separate component class. Add a dedicated class only if it develops meaningful runtime behavior that configuration cannot express.

### Items that should exit or not be created

- Deprecate the three legacy `LLMBase` classes named `*VisionModel` after compatibility aliases are available.
- Replace the current generic `FalVLM` with endpoint-family implementations whose schemas can be validated.
- Do not treat hard-coded `allowed_models` lists as authoritative. Use a provider `_get_models()` hook plus a small tested fallback list.
- Do not create checkpoint/version classes such as `Qwen2_5VLVLM`, `LLaVA15VLM`, or `Pixtral12BVLM`. Classes should represent stable provider/runtime behavior; `name` selects the model.
- Do not create a separate `IImageCaption`; captioning remains a VLM capability.
- BLIP-2, older LLaVA, OFA, and similar historical implementations should remain research or regression baselines rather than first production packages.

### Recommended implementation order

1. Harden `IPredictVision` and `VLMBase`, including normalized multimodal message parts and capability flags.
2. Add `IStreamingVLM`, `OpenAICompatibleVLM`, and `TransformersVLM`.
3. Migrate and test `GroqVLM`; deprecate `GroqVisionModel`.
4. Implement `GeminiVLM`, `FlorenceVLM`, and the general local Transformers path.
5. Add `OpenAIVLM` and `AnthropicVLM`.
6. Repair/migrate `HyperbolicVLM` and replace the generic Fal implementation with validated endpoint-family classes.
7. Add `QwenVLM` only for behavior beyond the generic Transformers adapter.

This first wave provides managed general vision, local general vision, and specialized deterministic vision without creating a package for every model checkpoint.

### Primary references

- Hugging Face image-text-to-text/VLM contract and local runtime examples: https://huggingface.co/docs/transformers/main/tasks/image_text_to_text
- Gemini image understanding capabilities: https://ai.google.dev/gemini-api/docs/image-understanding
- OpenAI multimodal Responses input schema: https://platform.openai.com/docs/api-reference/responses-streaming/response/file_search_call?lang=javascript
- Claude image sources and multi-image conversations: https://platform.claude.com/docs/en/build-with-claude/vision
- Mistral vision through Chat Completions: https://docs.mistral.ai/studio-api/conversations/vision
- Groq supported vision models and limits: https://console.groq.com/docs/vision
- Florence-2 task coverage and model card: https://huggingface.co/microsoft/Florence-2-large
- Qwen3-VL model family example: https://huggingface.co/Qwen/Qwen3-VL-30B-A3B-Instruct
